### ISSUE TITLE
GCN used for regression(predicting the node score)

### ISSUE THREAD

**Issue by anilesec** [tkipf/gcn#130]
Hi,

I used this GCN for predicting the nodes score. The mse is loss not reducing much. It reduced from 0.2 to 0.1963. I think it is still possible to reduce, I tried adding another conv layer, but nothing improved. Could you please share your opinion on this.

Log: [https://app.wandb.ai/aniles-ec/Network1/runs/dakqkd92?workspace=user-aniles-ec]

Thanks
Anil

**Comment by tkipf** [comment]
Regression with GCNs using mean squared error is very reasonable. GCNs
should be able to outperform an MLP applied to each node independently. How
does such an MLP perform in your case?

**Comment by anilesec** [comment]
"_MLP applied to each node independently_", do you mean that if we have graph of 20 nodes with labels on each node then consider these 20 nodes as different training sample and train an MLP to predict the node labels using MSE loss?

**Comment by tkipf** [comment]
Yes

**Comment by anilesec** [comment]
yes, I just implemented. For a graph of 20 nodes(4 feature on each node), I trained two layer MLP(hidden dim=128, relu non-linearity) with MSE loss. MSE Loss reduced from 0.05 to 0.03 for 1000 epochs. I used only one graph isntance to train this. 

From this, GCN seem to be underperforminng, not able to figure out what is the problem,

**Comment by tkipf** [comment]
Is your GCN model overfitting or underfitting?

**Comment by anilesec** [comment]
This is the situation:
Over 100 epochs:
Train loss reduced from 0.2291 to 0.1962(almost plateau at around 25th epoch)
Val loss reduced from 0.2033 to 0.1972(its not smooth and it always remains higher than train loss)

Run details and plots: https://app.wandb.ai/aniles-ec/Network1/runs/5134bu7f?workspace=user-aniles-ec
 
So, i think its neither overfitting nor underfitting.

**Comment by tkipf** [comment]
Sounds like underfitting to me if the training loss is ~5 times higher than
with using the MLP? Maybe try some other GNN architectures, like in this
paper: https://arxiv.org/abs/1802.04687

**Comment by anilesec** [comment]
Okay, thanks for suggestion. I forgot to mention that my graph is Fully connected(20 nodes), is this could be the reason, because all nodes might get the same representations. In effect, my predicted values for each node would be same. So model may not be learning. What do you think?

**Comment by tkipf** [comment]
Yes if your graph is fully-connected then you obtain the same feature
representation for each node, which will not work if you want to classify
nodes differently.

**Comment by anilesec** [comment]
so, that means I cannot use GCN, I would have to try Graph Attention networks or some other model which gives different representation?

**Comment by tkipf** [comment]
If you have a fully-connected graph, then you can simply use the
Transformer architecture from https://arxiv.org/abs/1706.03762 . No need to
use Graph Attention Networks (which are just Transformers with masked-out
attention for non-existing edges).

**Comment by tkipf** [comment]
You can find a TensorFlow implementation of Transformer-style
self-attention here:
https://github.com/deepmind/graph_nets/blob/master/graph_nets/modules.py#L593

**Comment by anilesec** [comment]
Thanks for suggestion! I did use that but it does not generalize for different distribution(for eg. model trained with uniform distribution perform poorly on multi-variate gaussian distribution)

**Comment by mo7amed7assan1911** [comment]
Hi, any one here :)

**Comment by mo7amed7assan1911** [comment]
I face the same problem as all my nodes get the same value after regression. But, my graphs not fully connected. How I could handle that, my model is:
 Hi, I try to make a node regression but I face a problem is that my model get the same value for all the nodes in the batch. The loss decreases but I think that is because it tries to get a middle value between all values in the targets. my model is:
class GCNRegressor(torch.nn.Module):
    def init(self, in_channels, out_channels):
        super(GCNRegressor, self).init()
        self.conv1 = GCNConv(in_channels, 16)
        self.conv2 = GCNConv(16, 32)
        self.conv3 = GCNConv(32, 64)
        self.fc = nn.Linear(64, out_channels)
    def forward(self, x, edge_index):
        x = F.relu(self.conv1(x, edge_index))
        x = F.relu(self.conv2(x, edge_index))
        x = F.relu(self.conv3(x, edge_index))
        x = self.fc(x)
        return x.squeeze(1)
